### PR TITLE
Accept and default to base64_urlsafe CSRF tokens.

### PR DESCRIPTION
**Pull request by dragonsinth**
Strict base64 is an onerous requirement, and moreover makes CSRF values inherently not web-safe.  Use base64url instead, and accept either encoding.

### CODE REVIEW

**Comment by jkingdon** [comment]
See https://github.com/ruby/ruby/pull/815 for more discussion on the behaviors of Base64.urlsafe_*

As for the rest of this pull request, could use some tests but seems reasonable enough to me.

**Comment by NZKoz** [comment]
The value is automatically escaped and completely transparent to users, why would we change this?

**Comment by dragonsinth** [comment]
There's two separate issues.  One on the validation side, and one on the production side.

1) On the validation side, it's to support interop.  Some domains use a mix of servers to handle various aspects of the site.  Some parts might be various different versions of rails.  Other parts might be Java, Go, or Node.  CSRF tokens (and session cookies in general) are not specific to any particular backend, but apply across the whole site.

Previous versions of Rails merely required a *string match* when comparing CSRF values on the validation side.  Requiring CSRF tokens to be strict base64 is onerous for interoperability, it requires a multi-framework backend site to update every participating framework to deal only in strict base64.  This seems bad given that there's no web standard for CSRF to be strict base64.

2) On the production side, urlsafe base64 is clearly preferable to strict base64, since it's already inherently websafe.  Consider the situation where a site uses a static HTML page and heavy JS / XHR.  How does the javascript application know what the right CSRF value is?  You can't put it in a meta tag on the page, because then you'd have to generate the page dynamically and could not use a static page anymore.  So instead you typically put it in a client-readable cookie.  The JS application reads the cookie value, and uses that value to feed into the X-Csrf-Token header.  The server ignores the client-readable cookie in favor of the http-only session cookie, and compares the CSRF embedded in the session cookie with the header value.

In this scenario, it's highly preferable for the CSRF token to be inherently websafe, since it has to be transported to the client in a cookie.  Reading cookie values client side has weak standards, some frameworks automatically url-decode the values, but others do not.  If the CSRF value is inherently websafe, this problem is avoided, and you don't run into the problem where the client-readable CSRF value is no longer string == the embedded session CSRF cookie unless you properly decode it.

**Comment by NZKoz** [comment]
On point 1, I don't think that changing the encoding alters that at all, the real problem is that 'malformed' tokens should probably just fall back to string compare rather than assuming it's invalid.  You could investigate doing that.

For the second point, I can see why it's annoying in some circumstances, but fix the client libraries that don't behave consistently rather than ignoring the issue and changing what we generate?

**Comment by dragonsinth** [comment]
Good point regarding point 1; falling back to string compare seems like a win-win all around.

On the production side, I suppose my argument boils down to: given that there are no defined standards (and therefore we have a somewhat arbitrary choice), why not choose the most portable, easiest to use encoding?

**Comment by dragonsinth** [comment]
Pushed a commit with changes to demo allowing string match.  What do you think?

**Comment by dragonsinth** [comment]
Hi, any chance of this going forward?  I added a commit to fall back to string comparison.

**Comment by dragonsinth** [comment]
FWIW, ruby-core just merged a patch to improve support for lenient decoding.

**Comment by sgrif** [comment]
@dragonsinth It seems like this is no longer needed, now that https://github.com/ruby/ruby/pull/815 has been merged. Is that correct?

**Comment by dragonsinth** [comment]
No, this PR would still be extremely useful.  The Ruby change is merely useful at providing the core methods for working with urlsafe base64.  Defining truly portable urlsafe CSRF tokens is needed in rails itself.

**Comment by nerdrew** [comment]
Will there be any movement on this?

**Comment by dragonsinth** [comment]
ping

**Comment by dragonsinth** [comment]
ping

File: actionpack/lib/action_controller/metal/request_forgery_protection.rb [hunk 1]
```diff
@@ -311,9 +312,17 @@ def compare_with_real_token(token, session)
         ActiveSupport::SecurityUtils.secure_compare(token, real_csrf_token(session))
       end
 
+      # String compare raw CSRF values, both as a fast-path and for interop with other systems.
+      def raw_compare_with_real_token(raw_token, session)
+        ActiveSupport::SecurityUtils.secure_compare(raw_token, raw_csrf_token(session))
+      end
```
Comment (jeremy) [inline actionpack/lib/action_controller/metal/request_forgery_protection.rb]: Out of scope for this PR?

**Review by jeremy — COMMENTED** [review]
Looks reasonable @dragonsinth. Pending test coverage demonstrating backward compatibility and lenience.

File: actionpack/lib/action_controller/metal/request_forgery_protection.rb [hunk 2]
```diff
@@ -329,5 +338,15 @@ def form_authenticity_param
       def protect_against_forgery?
         allow_forgery_protection
       end
+
+      # Base64.urlsafe_decode64 not to spec
+      def lenient_decode64(token)
+        Base64.urlsafe_decode64(token.ljust((token.length + 3) & ~3, '='))
+      end
```
[the remark this answers is not in the document]
Comment (dragonsinth) [reply actionpack/lib/action_controller/metal/request_forgery_protection.rb]: It was intended to make trailing `=` optional -- the current code requires them, and (at the time I wrote this) `Base64.urlsafe_decode64` also required them.

**Review by dragonsinth — COMMENTED** [review]
I need to get a rails environment setup to hack on this-- I spent a couple hours last night trying to get my mac env setup to be able to revisit this.  I couldn't get IJ setup right, but I might have gotten the command line working.

What's the formulation for running a test (like `actionpack/test/controller/request_forgery_protection_test.rb`) from command line?

File: actionpack/lib/action_controller/metal/request_forgery_protection.rb [hunk 3]
```diff
@@ -287,17 +288,17 @@ def valid_authenticity_token?(session, encoded_masked_token)
...
+          compare_with_real_token csrf_bytes, session
 
-        elsif masked_token.length == AUTHENTICITY_TOKEN_LENGTH * 2
+        elsif csrf_bytes.length == AUTHENTICITY_TOKEN_LENGTH * 2
           # Split the token into the one-time pad and the encrypted
           # value and decrypt it
-          one_time_pad = masked_token[0...AUTHENTICITY_TOKEN_LENGTH]
-          encrypted_csrf_token = masked_token[AUTHENTICITY_TOKEN_LENGTH..-1]
+          one_time_pad = csrf_bytes[0...AUTHENTICITY_TOKEN_LENGTH]
+          encrypted_csrf_token = csrf_bytes[AUTHENTICITY_TOKEN_LENGTH..-1]
```
[the remark this answers is not in the document]
Comment (jeremy) [reply actionpack/lib/action_controller/metal/request_forgery_protection.rb]: > 5 year old memories

😅🙏

File: actionpack/lib/action_controller/metal/request_forgery_protection.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply actionpack/lib/action_controller/metal/request_forgery_protection.rb]: Cool. Let's reflect that consideration in the comment, replacing "not to spec."

File: actionpack/lib/action_controller/metal/request_forgery_protection.rb (hunk 1 above)
Comment (jeremy) [reply actionpack/lib/action_controller/metal/request_forgery_protection.rb]: > but not so much on the decoding side-- it's kind of irrelevant whether or not you can decode a previously encoded token as a particular format-- it just has to match.

That's distinct from this change, and it bears closer consideration of the implications. It's buried here alongside base64url encoding.

Comment (dragonsinth) [reply actionpack/lib/action_controller/metal/request_forgery_protection.rb]: Fair enough! I'll focus on the base64 stuff.

File: actionpack/lib/action_controller/metal/request_forgery_protection.rb [hunk 4]
```diff
@@ -329,5 +338,15 @@ def form_authenticity_param
...
       end
+
+      # Base64.urlsafe_decode64 not to spec
+      def lenient_decode64(token)
+        Base64.urlsafe_decode64(token.ljust((token.length + 3) & ~3, '='))
+      end
+
+      # Base64.urlsafe_encode64 not to spec
+      def encode64(bin)
+        Base64.urlsafe_encode64(bin).tr('=', '')
```
[the remark this answers is not in the document]
Comment (dragonsinth) [reply actionpack/lib/action_controller/metal/request_forgery_protection.rb]: OH LOL!  I guess my ruby PR got merged 😂 
https://github.com/ruby/ruby/pull/815

`padding` wasn't an arg when I wrote this PR.

**Comment by dragonsinth** [comment]
@jeremy I pushed an update if you'd like to take another look

**Review by jeremy — COMMENTED** [review]
Can we demonstrate backward compatibility (with `strict_encode64` tokens) with test coverage as well?

**Comment by dragonsinth** [comment]
> Can we demonstrate backward compatibility (with `strict_encode64` tokens) with test coverage as well?

I was thinking that!  I tried digging into the test cases though, and I had a really hard time making sense of them. :/  Would love some suggestions on this.

**Comment by dragonsinth** [comment]
An isolated test case that just demonstrates that base64 strict encode can be lenient urlsafe decoded would be easy enough; I'm just not sure how to write something that ties into the test infra in a sensible way and covers the end-to-end flow.

**Comment by jeremy** [comment]
Give it a shot, using the existing suite as a guide!

**Comment by dragonsinth** [comment]
> Give it a shot, using the existing suite as a guide!

Aight, see what you think.

**Review by jeremy — APPROVED** [review]
Looks good @dragonsinth! Just needs a changelog entry. Your commentary in https://github.com/rails/rails/pull/18496#issuecomment-70557391 is a good starting point to explain to Rails users what is changed, why it matters, and how it affects their apps.

**Comment by dragonsinth** [comment]
@jeremy added some words to the changelog, see what you think.  I omitted all the stuff about interop-- that was more relevant to the idea of allowing string-matched CSRF, which we pulled out of this PR.

**Comment by dragonsinth** [comment]
(Note: PR title changed)

**Comment by dragonsinth** [comment]
(Also enabled "allow commits from maintainers")

**Comment by dragonsinth** [comment]
Awesome, thanks @jeremy !

**Comment by etiennebarrie** [comment]
While this is backwards compatible, it needs to also be forwards compatible, because when you deploy new code without downtime, servers with the new code will start emitting base64 url safe tokens which won't be valid for the servers running the old code. Let me know if you're not interested I'll take a look.

One other thing, the changelog is a bit misleading, it's not that it defaults to base64 url safe tokens, but you don't really have a choice in the matter, which is fine IMO, but it's not a default if you don't have an option.

**Comment by etiennebarrie** [comment]
Oh I didn't even realize but the test generates by luck a token that doesn't include any of the characters that differ between the encodings.

And as I just mentioned in #39076, even if it did, the test still pass, but not because we explicitly handle strict encoding, but because `Base64.urlsafe_decode64` also accepts strict encoded data, thanks to its implementation that converts to strict instead of directly decoding.

**Comment by dragonsinth** [comment]
@etiennebarrie Cool!  Thank you for following up on getting this rolled out safely.

**Pull request merged**